include/continuable/detail/other/coroutines.hpp

```diff
@@ -117,16 +121,24 @@ class awaitable {
 
   /// Suspend the current context
   // TODO Convert this to an r-value function once possible
-  void await_suspend(coroutine_handle<> h) {
+  coroutine_handle<> await_suspend(coroutine_handle<> h) {
     assert(result_.is_empty());
     // Forward every result to the current awaitable
     std::move(continuable_)
         .next([h, this](auto&&... args) mutable {
           assert(result_.is_empty());
           result_ = result_t::from(std::forward<decltype(args)>(args)...);
-          h.resume();
+
+          // If nullptr, it means that the promise was not resolved
+          // immediately (i.e., the awaitable await_suspend method
+          // has already returned). That means we must call the
+          // resume function from the continuation chain.
+          if(handle_.exchange(h) == nullptr) {
```

Comment:
Ì would prefer checking the validity of the the coroutine_handle through operator bool.

Reply:
Also you can relax the memory order to handle_.exchange(h, std::memory_order_acq_rel)